From d3e925b70aa3b1c73328029d31a514fee44c867a Mon Sep 17 00:00:00 2001
From: WofWca <wofwca@protonmail.com>
Date: Sun, 9 Nov 2025 21:00:17 +0400
Subject: [PATCH] improvement: gibs: keep player angles

The previous commit added proper initial positions
to each individual gib.
And now it's time to also make sure
that they're facing the right direction.
---
 code/cgame/cg_effects.c   | 88 ++++++++++++++++++++++-----------------
 code/cgame/cg_event.c     |  2 +-
 code/cgame/cg_local.h     |  2 +-
 code/cgame/cg_localents.c |  6 ++-
 4 files changed, 57 insertions(+), 41 deletions(-)

diff --git a/code/cgame/cg_effects.c b/code/cgame/cg_effects.c
index ab6bb416..09986505 100644
--- a/code/cgame/cg_effects.c
+++ b/code/cgame/cg_effects.c
@@ -526,7 +526,8 @@ void CG_Bleed( vec3_t origin, int entityNum ) {
 CG_LaunchGib
 ==================
 */
-void CG_LaunchGib( vec3_t origin, vec3_t velocity, qhandle_t hModel ) {
+void CG_LaunchGib( const vec3_t origin, const vec3_t angles,
+					const vec3_t velocity, const qhandle_t hModel ) {
 	localEntity_t	*le;
 	refEntity_t		*re;
 
@@ -538,7 +539,7 @@ void CG_LaunchGib( vec3_t origin, vec3_t velocity, qhandle_t hModel ) {
 	le->endTime = le->startTime + 5000 + random() * 3000;
 
 	VectorCopy( origin, re->origin );
-	AxisCopy( axisDefault, re->axis );
+	AnglesToAxis( angles, re->axis );
 	re->hModel = hModel;
 
 	le->pos.trType = TR_GRAVITY;
@@ -561,8 +562,12 @@ Generated a bunch of gibs launching out from the bodies location
 */
 #define	GIB_VELOCITY	250
 #define	GIB_JUMP		250
-void CG_GibPlayer( vec3_t playerOrigin ) {
+void CG_GibPlayer( const vec3_t playerOrigin, const vec3_t playerAngles ) {
 	vec3_t	origin, velocity;
+	// Generally only the head should have pitch,
+	// the rest of the body is upright.
+	vec3_t	bodyAngles;
+	vec3_t	forward, right, up;
 	// See `playerMins`, `playerMaxs`.
 	// TODO we could try to check the actual `mins` and `maxs`
 	// (do we have them available on the client though?),
@@ -575,15 +580,20 @@ void CG_GibPlayer( vec3_t playerOrigin ) {
 		return;
 	}
 
+	VectorCopy( playerAngles, bodyAngles );
+	// This way `up` is always `{0,0,1}`.
+	bodyAngles[PITCH] = 0;
+	AngleVectors( bodyAngles, forward, right, up );
+
 	VectorCopy( playerOrigin, origin );
 	origin[2] = bottom + 0.95 * playerHeight;
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
 	if ( rand() & 1 ) {
-		CG_LaunchGib( origin, velocity, cgs.media.gibSkull );
+		CG_LaunchGib( origin, playerAngles, velocity, cgs.media.gibSkull );
 	} else {
-		CG_LaunchGib( origin, velocity, cgs.media.gibBrain );
+		CG_LaunchGib( origin, playerAngles, velocity, cgs.media.gibBrain );
 	}
 
 	// allow gibs to be turned off for speed
@@ -596,91 +606,93 @@ void CG_GibPlayer( vec3_t playerOrigin ) {
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
-	CG_LaunchGib( origin, velocity, cgs.media.gibAbdomen );
+	CG_LaunchGib( origin, bodyAngles, velocity, cgs.media.gibAbdomen );
 
 	VectorCopy( playerOrigin, origin );
 	origin[2] = bottom + 0.78 * playerHeight;
-	origin[1] += 0.8 * playerRadius;
-	origin[0] += 0.3 * playerRadius;
+	VectorMA( origin, 0.8 * playerRadius, right, origin );
+	VectorMA( origin, 0.3 * playerRadius, forward, origin );
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
-	CG_LaunchGib( origin, velocity, cgs.media.gibArm );
+	CG_LaunchGib( origin, bodyAngles, velocity, cgs.media.gibArm );
 
 	VectorCopy( playerOrigin, origin );
 	origin[2] = bottom + 0.80 * playerHeight;
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
-	CG_LaunchGib( origin, velocity, cgs.media.gibChest );
+	CG_LaunchGib( origin, bodyAngles, velocity, cgs.media.gibChest );
 
 	VectorCopy( playerOrigin, origin );
 	origin[2] = bottom + 0.66 * playerHeight;
-	origin[1] += 0.8 * playerRadius;
-	origin[0] += 0.2 * playerRadius;
+	VectorMA( origin, 0.8 * playerRadius, right, origin );
+	VectorMA( origin, 0.2 * playerRadius, forward, origin );
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
-	CG_LaunchGib( origin, velocity, cgs.media.gibFist );
+	CG_LaunchGib( origin, bodyAngles, velocity, cgs.media.gibFist );
 
 	VectorCopy( playerOrigin, origin );
 	origin[2] = bottom + 0.05 * playerHeight;
-	origin[1] += -0.5 * playerRadius;
-	origin[0] += -0.5 * playerRadius;
+	VectorMA( origin, -0.5 * playerRadius, right, origin );
+	VectorMA( origin, -0.5 * playerRadius, forward, origin );
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
-	CG_LaunchGib( origin, velocity, cgs.media.gibFoot );
+	CG_LaunchGib( origin, bodyAngles, velocity, cgs.media.gibFoot );
 
 	VectorCopy( playerOrigin, origin );
 	origin[2] = bottom + 0.65 * playerHeight;
-	origin[1] += -0.6 * playerRadius;
-	origin[0] += -0.2 * playerRadius;
+	VectorMA( origin, -0.6 * playerRadius, right, origin );
+	VectorMA( origin, -0.2 * playerRadius, forward, origin );
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
-	CG_LaunchGib( origin, velocity, cgs.media.gibForearm );
+	CG_LaunchGib( origin, bodyAngles, velocity, cgs.media.gibForearm );
 
 	VectorCopy( playerOrigin, origin );
 	origin[2] = bottom + 0.57 * playerHeight;
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
-	CG_LaunchGib( origin, velocity, cgs.media.gibIntestine );
+	CG_LaunchGib( origin, bodyAngles, velocity, cgs.media.gibIntestine );
 
 	VectorCopy( playerOrigin, origin );
 	origin[2] = bottom + 0.42 * playerHeight;
-	origin[1] += 0.5 * playerRadius;
-	origin[0] += 0.1 * playerRadius;
+	VectorMA( origin, 0.5 * playerRadius, right, origin );
+	VectorMA( origin, 0.1 * playerRadius, forward, origin );
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
-	CG_LaunchGib( origin, velocity, cgs.media.gibLeg );
+	CG_LaunchGib( origin, bodyAngles, velocity, cgs.media.gibLeg );
 
 	VectorCopy( playerOrigin, origin );
 	origin[2] = bottom + 0.44 * playerHeight;
-	origin[1] += -0.5 * playerRadius;
-	origin[0] += -0.2 * playerRadius;
+	VectorMA( origin, -0.5 * playerRadius, right, origin );
+	VectorMA( origin, -0.2 * playerRadius, forward, origin );
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
-	CG_LaunchGib( origin, velocity, cgs.media.gibLeg );
+	CG_LaunchGib( origin, bodyAngles, velocity, cgs.media.gibLeg );
 }
 void CG_GibPlayerOld( vec3_t playerOrigin ) {
-	vec3_t	origin, velocity;
+	vec3_t	origin, angles, velocity;
 
 	if ( !cg_blood.integer ) {
 		return;
 	}
 
+	VectorClear(angles);
+
 	VectorCopy( playerOrigin, origin );
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
 	if ( rand() & 1 ) {
-		CG_LaunchGib( origin, velocity, cgs.media.gibSkull );
+		CG_LaunchGib( origin, angles, velocity, cgs.media.gibSkull );
 	} else {
-		CG_LaunchGib( origin, velocity, cgs.media.gibBrain );
+		CG_LaunchGib( origin, angles, velocity, cgs.media.gibBrain );
 	}
 
 	// allow gibs to be turned off for speed
@@ -692,55 +704,55 @@ void CG_GibPlayerOld( vec3_t playerOrigin ) {
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
-	CG_LaunchGib( origin, velocity, cgs.media.gibAbdomen );
+	CG_LaunchGib( origin, angles, velocity, cgs.media.gibAbdomen );
 
 	VectorCopy( playerOrigin, origin );
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
-	CG_LaunchGib( origin, velocity, cgs.media.gibArm );
+	CG_LaunchGib( origin, angles, velocity, cgs.media.gibArm );
 
 	VectorCopy( playerOrigin, origin );
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
-	CG_LaunchGib( origin, velocity, cgs.media.gibChest );
+	CG_LaunchGib( origin, angles, velocity, cgs.media.gibChest );
 
 	VectorCopy( playerOrigin, origin );
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
-	CG_LaunchGib( origin, velocity, cgs.media.gibFist );
+	CG_LaunchGib( origin, angles, velocity, cgs.media.gibFist );
 
 	VectorCopy( playerOrigin, origin );
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
-	CG_LaunchGib( origin, velocity, cgs.media.gibFoot );
+	CG_LaunchGib( origin, angles, velocity, cgs.media.gibFoot );
 
 	VectorCopy( playerOrigin, origin );
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
-	CG_LaunchGib( origin, velocity, cgs.media.gibForearm );
+	CG_LaunchGib( origin, angles, velocity, cgs.media.gibForearm );
 
 	VectorCopy( playerOrigin, origin );
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
-	CG_LaunchGib( origin, velocity, cgs.media.gibIntestine );
+	CG_LaunchGib( origin, angles, velocity, cgs.media.gibIntestine );
 
 	VectorCopy( playerOrigin, origin );
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
-	CG_LaunchGib( origin, velocity, cgs.media.gibLeg );
+	CG_LaunchGib( origin, angles, velocity, cgs.media.gibLeg );
 
 	VectorCopy( playerOrigin, origin );
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
-	CG_LaunchGib( origin, velocity, cgs.media.gibLeg );
+	CG_LaunchGib( origin, angles, velocity, cgs.media.gibLeg );
 }
 
 /*
diff --git a/code/cgame/cg_event.c b/code/cgame/cg_event.c
index 4191888d..83a9cb58 100644
--- a/code/cgame/cg_event.c
+++ b/code/cgame/cg_event.c
@@ -1222,7 +1222,7 @@ void CG_EntityEvent( centity_t *cent, vec3_t position ) {
 		if (cg_oldGibs.integer) {
 			CG_GibPlayerOld( cent->lerpOrigin );
 		} else {
-			CG_GibPlayer( cent->lerpOrigin );
+			CG_GibPlayer( cent->lerpOrigin, cent->lerpAngles );
 		}
 		break;
 
diff --git a/code/cgame/cg_local.h b/code/cgame/cg_local.h
index 4a4da431..e87bb14f 100644
--- a/code/cgame/cg_local.h
+++ b/code/cgame/cg_local.h
@@ -1418,7 +1418,7 @@ void CG_LightningBoltBeam( vec3_t start, vec3_t end );
 #endif
 void CG_ScorePlum( int client, vec3_t org, int score );
 
-void CG_GibPlayer( vec3_t playerOrigin );
+void CG_GibPlayer( const vec3_t playerOrigin, const vec3_t playerAngles );
 void CG_GibPlayerOld( vec3_t playerOrigin );
 void CG_BigExplode( vec3_t playerOrigin );
 
diff --git a/code/cgame/cg_localents.c b/code/cgame/cg_localents.c
index d0ddef2c..c1683908 100644
--- a/code/cgame/cg_localents.c
+++ b/code/cgame/cg_localents.c
@@ -690,7 +690,11 @@ void CG_AddInvulnerabilityJuiced( localEntity_t *le ) {
 		if (cg_oldGibs.integer) {
 			CG_GibPlayerOld( le->refEntity.origin );
 		} else {
-			CG_GibPlayer( le->refEntity.origin );
+			vec3_t angles;
+			// Angles don't matter much here.
+			VectorClear( angles );
+
+			CG_GibPlayer( le->refEntity.origin, angles );
 		}
 	}
 	else {
-- 
2.54.0

